### PR TITLE
docker_25: init at 25.0.3

### PR DESCRIPTION
**Pull request by jmbaur**
## Description of changes

Release notes:
 - https://github.com/moby/moby/releases/tag/v25.0.0
 - https://github.com/moby/moby/releases/tag/v25.0.1
 - https://github.com/moby/moby/releases/tag/v25.0.2
 - https://github.com/moby/moby/releases/tag/v25.0.3

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -31125,7 +31125,7 @@ with pkgs;
   dnglab = callPackage ../tools/graphics/dnglab { };
 
   inherit (callPackage ../applications/virtualization/docker {})
-    docker_20_10 docker_24;
+    docker_20_10 docker_24 docker_25_rc1;
```
[the remark this answers is not in the document]
Comment (jmbaur) [reply pkgs/top-level/all-packages.nix]: makes sense! Thanks :)

**Review by emilylange — DISMISSED** [review]
Cross posting https://github.com/NixOS/nixpkgs/pull/278531#discussion_r1451611629 because of the change to `LimitNOFILE = 1048576;` made in b854c8aef57c870dae54f649acbd90e7d0de22f8.

**Comment by jmbaur** [comment]
> Cross posting [#278531 (comment)](https://github.com/NixOS/nixpkgs/pull/278531#discussion_r1451611629) because of the change to `LimitNOFILE = 1048576;` made in [b854c8a](https://github.com/NixOS/nixpkgs/commit/b854c8aef57c870dae54f649acbd90e7d0de22f8).

The reasoning in https://github.com/moby/moby/pull/45534#issue-1708467087 sounds good to me in favor of using systemd's defaults. Removing the explicit setting of `LimitNOFILE` in this module.

**Comment by emilylange** [comment]
@infinisil is the failing `pkgs/by-name` check due to the `inherit (callPackage` pattern a false positive or intended?

I did not immediately spot documentation for that `inherit` pattern in the `pkgs/by-name` docs.
Though I might have missed something (on mobile rn).

**Comment by infinisil** [comment]
> @infinisil is the failing `pkgs/by-name` check due to the `inherit (callPackage` pattern a false positive or intended?

False positive, please ignore, master won't break from merging like this! Am working on a fix for this :)

**Comment by infinisil** [comment]
> Am working on a fix for this :)

Will be fixed with https://github.com/NixOS/nixpkgs/pull/285089

**Comment by ereslibre** [comment]
Result of `nixpkgs-review pr 278607` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python312Packages.jupyter-repo2docker</li>
    <li>python312Packages.jupyter-repo2docker.dist</li>
  </ul>
</details>
<details>
  <summary>20 packages built:</summary>
  <ul>
    <li>airlift</li>
    <li>airlift.dist</li>
    <li>apptainer</li>
    <li>apptainer-overriden-nixos</li>
    <li>charliecloud</li>
    <li>docker (docker_24)</li>
    <li>docker-gc</li>
    <li>docker-sbom</li>
    <li>docker_20_10</li>
    <li>docker_20_10.man</li>
    <li>docker_25</li>
    <li>fn-cli</li>
    <li>nvidia-docker</li>
    <li>pipework</li>
    <li>python311Packages.jupyter-repo2docker</li>
    <li>python311Packages.jupyter-repo2docker.dist</li>
    <li>singularity</li>
    <li>singularity-overriden-nixos</li>
    <li>udocker</li>
    <li>udocker.dist</li>
  </ul>
</details>

**Comment by infinisil** [comment]
The PR to fix the `pkgs/by-name` check is now merged and used in CI, so it won't fail anymore the next time you push :)

**Comment by jmbaur** [comment]
Thanks @infinisil :)

File: pkgs/applications/virtualization/docker/default.nix [hunk 2]
```diff
@@ -306,4 +306,18 @@ rec {
     tiniRev = "v0.19.0";
     tiniHash = "sha256-ZDKu/8yE5G0RYFJdhgmCdN3obJNyRWv6K/Gd17zc1sI=";
   };
+
+  docker_25 = callPackage dockerGen rec {
+    version = "25.0.2";
+    cliRev = "v${version}";
+    cliHash = "sha256-tzPuehqV1iK9F2Hc3XigxctHEWzal0BoO7Qf8FOdX8Y=";
+    mobyRev = "v${version}";
+    mobyHash = "sha256-Q3he5WuybWQ9ScTPbZYbKk6KXN9PWUT7zFI2OFdnfSw=";
```
Comment (milas) [inline pkgs/applications/virtualization/docker/default.nix:315]:
```suggestion
    version = "25.0.3";
    cliRev = "v${version}";
    cliHash = "sha256-Jvb0plV1O/UzrcpzN4zH5OulmTVF+p9UQQQ9xqkiObQ=";
    mobyRev = "v${version}";
    mobyHash = "sha256-cDlRVdQNzH/X2SJUYHK1QLUHlKQtSyRYCVbz3wPx1ZM=";
```

**Review by milas — COMMENTED** [review]
Updated hashes for [v25.0.3](https://github.com/moby/moby/releases/tag/v25.0.3)

<sup>disclaimer: I'm a Docker Inc employee but I'm not actually a Moby maintainer</sup>

**Review by vdemeester — APPROVED** [review]
LGTM 🐯

**Comment by bhavitsharma** [comment]
would love to see this PR merged! we're waiting on some new OCI compliant layer remote cache features.

**Pull request merged**